storm-core/src/jvm/org/apache/storm/utils/Utils.java

```diff
@@ -1772,6 +1783,7 @@ public static String uuid() {
     public static void exitProcess (int val, String msg) {
         String combinedErrorMessage = "Halting process: " + msg;
         LOG.error(combinedErrorMessage, new RuntimeException(combinedErrorMessage));
+        isJVMShutdownInitiated.set(Boolean.TRUE);
```

Comment:
ShellSpout and ShellBolt just call System.exit() directly from their die() method. You may want to change those too.